pkg/resource/node-resource-manager.go

```diff
@@ -0,0 +1,233 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	predictionv1 "github.com/gocrane/api/pkg/generated/informers/externalversions/prediction/v1alpha1"
+	predictionlisters "github.com/gocrane/api/pkg/generated/listers/prediction/v1alpha1"
+	predictionapi "github.com/gocrane/api/prediction/v1alpha1"
+	"github.com/gocrane/crane/pkg/common"
+	"github.com/gocrane/crane/pkg/prediction/config"
+)
+
+const (
+	ExtResourcePrefixFormat = "ext-resource.node.gocrane.io/%s"
+	MinDeltaRatio           = 0.1
+	StateExpiration         = 1 * time.Minute
+	TspUpdateInterval       = 20 * time.Second
+)
+
+type NodeResourceManager struct {
+	nodeName string
+	client   clientset.Interface
+
+	nodeLister corelisters.NodeLister
+	nodeSynced cache.InformerSynced
+
+	tspLister predictionlisters.TimeSeriesPredictionLister
+	tspSynced cache.InformerSynced
+
+	stateChann chan map[string][]common.TimeSeries
+
+	//TODO: use state to be a backup of tsp
+	// A copy of data from stateChann
+	state map[string][]common.TimeSeries
+	// Updated when get new data from stateChann, used to determine whether state has expired
+	lastStateTime time.Time
+}
+
+func NewNodeResourceManager(client clientset.Interface, nodeName string, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer,
```

Comment:
we neead add metrics for the node resource manager

Reply:
ok